modules/apps/notification/notification-service/src/main/java/com/liferay/notification/internal/upgrade/v3_10_4/DeleteStaleNotificationUpgradeProcess.java

```diff
@@ -0,0 +1,244 @@
+
+				while (resultSet2.next()) {
+					_deleteNotificationRecipient(resultSet2.getLong(1));
+				}
+
+				Repository repository = _getRepository(companyId);
+
+				if (repository != null) {
+					try {
+						Folder folder = _portletFileRepository.getPortletFolder(
+							repository.getRepositoryId(),
+							DLFolderConstants.DEFAULT_PARENT_FOLDER_ID,
+							String.valueOf(notificationQueueEntryId));
+
+						_portletFileRepository.deletePortletFolder(
+							folder.getFolderId());
+					}
+					catch (PortalException portalException) {
+						if (_log.isDebugEnabled()) {
+							_log.debug(portalException);
+						}
+					}
+				}
+			}
+
+			deletePreparedStatement1.executeBatch();
+			deletePreparedStatement2.executeBatch();
+		}
+	}
+
+	private void _deleteNotificationRecipient(long notificationRecipientId)
+		throws Exception {
+
+		runSQL(
+			StringBundler.concat(
+				"delete from NotificationRecipient where ",
+				"notificationRecipientId = ", notificationRecipientId));
+		runSQL(
+			StringBundler.concat(
+				"delete from NotificationRecipientSetting where ",
+				"notificationRecipientId = ", notificationRecipientId));
+	}
+
+	private void _deleteNotificationTemplate() throws Exception {
+		try (PreparedStatement deletePreparedStatement1 =
+				 _getDeletePreparedStatement("NotificationTemplate",
+					 "notificationTemplateId");
+			 PreparedStatement deletePreparedStatement2 =
+				 _getDeletePreparedStatement("NTemplateAttachment",
+					 "notificationTemplateId");
+			 PreparedStatement preparedStatement1 =
+				 _getSelectPreparedStatement();
+			 PreparedStatement preparedStatement2 =
+				 _getSelectPreparedStatement("NotificationTemplate");
+
+			 ResultSet resultSet1 = preparedStatement2.executeQuery();
+			) {
+
+			while (resultSet1.next()) {
+				long notificationTemplateId = resultSet1.getLong(
+					"notificationTemplateId");
+
+				_deleteResourcePermissions(
+					NotificationTemplate.class.getName(),
+					notificationTemplateId, resultSet1.getLong("companyId"));
+
+				deletePreparedStatement1.setLong(1, notificationTemplateId);
+				deletePreparedStatement2.setLong(1, notificationTemplateId);
+
+				deletePreparedStatement1.addBatch();
+				deletePreparedStatement2.addBatch();
+
+				preparedStatement1.setLong(
+					1,
+					PortalUtil.getClassNameId(
+						NotificationTemplate.class.getName()));
+				preparedStatement1.setLong(2, notificationTemplateId);
+
+				ResultSet resultSet2 = preparedStatement1.executeQuery();
+
+				while (resultSet2.next()) {
+					_deleteNotificationRecipient(resultSet2.getLong(1));
+				}
+			}
+
+			deletePreparedStatement1.executeBatch();
+			deletePreparedStatement2.executeBatch();
+		}
+	}
+
+	private void _deleteResourcePermissions(
+			String className, long primarykey, long companyId)
+		throws Exception {
+
+		_resourcePermissionLocalService.deleteResourcePermissions(
+			companyId, className, ResourceConstants.SCOPE_INDIVIDUAL,
+			String.valueOf(primarykey));
+	}
```

Comment:
Sort the parameters following the order in `_resourcePermissionLocalService.deleteResourcePermissions()`:
`companyId > className > primarykey`